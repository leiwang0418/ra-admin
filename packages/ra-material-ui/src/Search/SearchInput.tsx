import React, { useState, useRef } from "react";
import PropTypes, { InferProps } from "prop-types";
import { defineMessages, useIntl } from "react-intl";
import Paper from "@material-ui/core/Paper";
import IconButton from "@material-ui/core/IconButton";
import SearchIcon from "@material-ui/icons/Search";
import Divider from "@material-ui/core/Divider";
import CancelRoundedIcon from "@material-ui/icons/CancelRounded";
import InputAdornment from "@material-ui/core/InputAdornment";
import InputBase from "@material-ui/core/InputBase";
import TuneIcon from "@material-ui/icons/Tune";
import { makeStyles } from "@material-ui/styles";
import { styled, alpha } from "@material-ui/core/styles";
import { Theme } from "@material-ui/core/styles";

const SearchRoot = styled(Paper)(({ theme }) => ({
	borderRadius: theme.shape.borderRadius,
	backgroundColor: alpha(theme.palette.common.black, 0.1),
	padding: "2px 4px",
	alignItems: "center",
}));

const useStyles = makeStyles((theme: Theme) => ({
	root: {
		boxShadow: theme.shadows[2],
		backgroundColor: `${theme.palette.common.white} !important`,
	},
}));

const SearchPropTypes = {
	placeholderLabel: PropTypes.object,
	clearLabel: PropTypes.object,
	searchLabel: PropTypes.object,
};

export type SearchTypes = InferProps<typeof SearchPropTypes>;

const SearchInput = (props: SearchTypes) => {
	const {
		placeholderLabel = messages.placeholder,
		clearLabel = messages.clearLabel,
		searchLabel = messages.searchLabel,
	} = props;

	const intl = useIntl();
	const [value, setValue] = useState("");
	const classes = useStyles();
	const [focused, setFocused] = useState(false);
	const inputEl = useRef<HTMLInputElement | null>(null);

	const CancelIcon = (
		<IconButton
			sx={{ visibility: value ? "visible" : "hidden" }}
			aria-label={clearLabel ? intl.formatMessage(clearLabel) : ""}
			onClick={() => {
				setValue("");
				inputEl.current!.focus();
			}}
		>
			<CancelRoundedIcon onMouseDown={(e) => e.preventDefault()} />
		</IconButton>
	);
	const DividerIcon = (
		<Divider
			sx={{
				height: 20,
				m: 0.5,
				visibility: value ? "visible" : "hidden",
			}}
			orientation="vertical"
		/>
	);
	const BasicSearch = (
		<InputAdornment position="start">
			<IconButton
				aria-label={searchLabel ? intl.formatMessage(searchLabel) : ""}
				onClick={() => console.log("basic search")}
			>
				<SearchIcon onMouseDown={(e) => e.preventDefault()} />
			</IconButton>
		</InputAdornment>
	);
	const AdvancedSearch = (
		<IconButton
			onClick={() => console.log("advanced search")}
			sx={{ p: "5px" }}
			aria-label={searchLabel ? intl.formatMessage(searchLabel) : ""}
		>
			<TuneIcon />
		</IconButton>
	);

	return (
		<SearchRoot
			className={focused ? classes.root : ""}
			elevation={focused ? 1 : 0}
		>
			<InputBase
				fullWidth
				startAdornment={BasicSearch}
				endAdornment={
					<React.Fragment>
						{CancelIcon}
						{DividerIcon}
						{AdvancedSearch}
					</React.Fragment>
				}
				onChange={(e) => setValue(e.target.value)}
				value={value}
				onFocus={() => setFocused(true)}
				onBlur={() => setFocused(false)}
				inputRef={inputEl}
				inputProps={{
					"aria-label": placeholderLabel
						? intl.formatMessage(placeholderLabel)
						: "",
					placeholder: placeholderLabel
						? intl.formatMessage(placeholderLabel)
						: "",
				}}
			/>
		</SearchRoot>
	);
};

SearchInput.propTypes = SearchPropTypes;

const messages = defineMessages({
	placeholder: {
		id: "ra.defaultSearch.placeholder",
		defaultMessage: "Search...",
	},
	clearLabel: {
		id: "ra.defaultSearch.clear.label",
		defaultMessage: "Clear",
	},
	searchLabel: {
		id: "ra.defaultSearch.search.label",
		defaultMessage: "Search",
	},
});

export default SearchInput;
