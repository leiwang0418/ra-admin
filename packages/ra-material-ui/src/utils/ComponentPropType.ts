import { isValidElementType } from "react-is";

const ComponentPropType = (
	props: { [key: string]: any },
	propName: string,
	componentName: string
) => {
	if (props[propName] && !isValidElementType(props[propName])) {
		return new Error(
			`Invalid prop '${propName}' supplied to '${componentName}': the prop is not a valid React component`
		);
	}
};

export default ComponentPropType;