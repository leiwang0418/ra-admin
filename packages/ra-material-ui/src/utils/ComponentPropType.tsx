import { ReactPropTypes } from "react";
import { isValidElementType } from "react-is";
import { Validator } from "prop-types";

export default (
	props: { [key: string]: any },
	propName: string,
	componentName: string
) => {
	const prop = props[propName];
	if (prop && !isValidElementType(prop)) {
		return new Error(
			`Invalid prop '${propName} supplied to '${componentName}': the prop is not a valid React component`
		);
	}
};