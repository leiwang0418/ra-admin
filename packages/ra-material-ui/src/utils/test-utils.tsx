import React, { FC } from "react";
import { render, RenderOptions } from "@testing-library/react";
import { IntlProvider } from "react-intl";
import en from "../compiled-lang/en.json";
import { ThemeProvider } from "@material-ui/styles";
import { createTheme } from "@material-ui/core/styles";

const theme = createTheme();

const customRender = (
	ui: React.ReactElement,
	{ locale = "en", ...renderOptions } = {}
) => {
	const Wrapper: FC = ({ children }) => (
		<IntlProvider locale={locale} messages={en}>
			<ThemeProvider theme={theme}>{children}</ThemeProvider>
		</IntlProvider>
	);

	return render(ui, { wrapper: Wrapper, ...renderOptions });
};

export * from "@testing-library/react";

export { customRender as render };
