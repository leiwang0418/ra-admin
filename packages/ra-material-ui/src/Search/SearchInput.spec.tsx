import React from "react";
import { render, screen } from "../utils/test-utils";
import SearchInput from "./SearchInput";

describe("snapshot test", () => {
	it("render the basic SearchInput", () => {
		const { container } = render(<SearchInput />);

		expect(container).toMatchSnapshot();

		expect(screen.getByLabelText(/Search.../i)).toBeInTheDocument();
	});
});
