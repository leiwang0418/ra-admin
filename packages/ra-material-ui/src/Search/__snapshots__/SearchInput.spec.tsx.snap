// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test render the basic SearchInput 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0  css-10umtpx-MuiPaper-root"
  >
    <div
      class="MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-adornedStart MuiInputBase-adornedEnd css-luel6v-MuiInputBase-root"
    >
      <div
        class="MuiInputAdornment-root MuiInputAdornment-positionStart css-ittuaa-MuiInputAdornment-root"
      >
        <span
          class="notranslate"
        >
          â€‹
        </span>
        <button
          aria-label="Search"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="SearchIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
      <input
        aria-label="Search..."
        class="MuiInputBase-input MuiInputBase-inputAdornedStart MuiInputBase-inputAdornedEnd css-yz9k0d-MuiInputBase-input"
        placeholder="Search..."
        type="text"
        value=""
      />
      <button
        aria-label="Clear"
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-1skso5k-MuiButtonBase-root-MuiIconButton-root"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="CancelRoundedIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 2C6.47 2 2 6.47 2 12s4.47 10 10 10 10-4.47 10-10S17.53 2 12 2zm4.3 14.3c-.39.39-1.02.39-1.41 0L12 13.41 9.11 16.3c-.39.39-1.02.39-1.41 0a.9959.9959 0 0 1 0-1.41L10.59 12 7.7 9.11a.9959.9959 0 0 1 0-1.41c.39-.39 1.02-.39 1.41 0L12 10.59l2.89-2.89c.39-.39 1.02-.39 1.41 0 .39.39.39 1.02 0 1.41L13.41 12l2.89 2.89c.38.38.38 1.02 0 1.41z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <hr
        class="MuiDivider-root MuiDivider-fullWidth MuiDivider-vertical css-1lxet3n-MuiDivider-root"
      />
      <button
        aria-label="Search"
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-14cutrz-MuiButtonBase-root-MuiIconButton-root"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="TuneIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M3 17v2h6v-2H3zM3 5v2h10V5H3zm10 16v-2h8v-2h-8v-2h-2v6h2zM7 9v2H3v2h4v2h2V9H7zm14 4v-2H11v2h10zm-6-4h2V7h4V5h-4V3h-2v6z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</div>
`;
